DOCSP-35966 query builder

ccho-mongodb opened the pull request:
<!--
Replace this notice by a description of your feature/bugfix.
This will help reviewers and should be a good start for the documentation.
-->

JIRA: https://jira.mongodb.org/browse/DOCSP-35966
Staging: https://preview-mongodbcchomongodb.gatsbyjs.io/laravel/DOCSP-35966-query-builder/query-builder/


### Checklist

- [ ] Add tests and ensure they pass
- [ ] Add an entry to the CHANGELOG.md file
- [ ] Update documentation for new features

Commit:
DOCSP-35966: Query builder standardization

Commit:
WIP

Commit:
WIP

Commit:
wip

Commit:
WIP

Commit:
WIP

Commit:
fix RST

Commit:
fix RST

Commit:
WIP

Commit:
WIP

Commit:
fix

Commit:
wip

Commit:
wip

Commit:
WIP

Commit:
WIP

Commit:
WIP

Commit:
fix indent

Commit:
fix RST

Commit:
format code

Commit:
reformat code

Commit:
WIP

Commit:
elemmatch update

Commit:
fixes

Commit:
capitalize

Commit:
single quotes

Commit:
standardize value formatting, fix link

GromNaN commented on docs/includes/query-builder/MovieController.php:
```diff
@@ -0,0 +1,691 @@
+use function print_r;
+use function view;
+
+class MovieController extends Controller
```
Could you write the examples in a TestCase like I'm doing in https://github.com/mongodb/laravel-mongodb/pull/2775

rustagir commented on docs/query-builder.txt:
```diff
@@ -9,588 +9,1095 @@ Query Builder
-other MongoDB specific operators/operations.
+In this guide, you can learn how to use the {+odm-short+} extension of
+the Laravel query builder to work with a MongoDB database. The query builder
+lets you write queries for any supported database by using the same fluent
+interface and syntax.
```
Q: What does "same" mean in this context?

ccho-mongodb replied:
I think I had "a common" previously to represent the "fluent interface and syntax" which might be more clear than "the same". Let me know if one is more clear than the other, or how you might otherwise word it.

rustagir replied:
maybe you can just add "... the same ... as the Laravel query builder"?

ccho-mongodb replied:
Thanks for that perspective!

I wanted to describe the query builder rather than comparing it with something else, but it seems like using noun phrases (e.g. https://github.com/mongodb/laravel-mongodb/pull/2790#discussion_r1538001671) can cause some confusion.
In this sentence, the items being compared are syntax you use to access each of the supported databases rather than something with the query builder (given it's describing the query builder).

I'll spend a few minutes to try to reorder the sentence in a way that avoids connecting noun phrases to restrictive clauses.

rustagir commented on docs/query-builder.txt:
```diff
@@ -9,588 +9,1095 @@ Query Builder
+Laravel provides a **facade** to access the query builder class ``DB``, which
+lets you perform database operations. Facades are static interfaces to
+classes that make the syntax more concise, avoid runtime errors, and improve
+testability.
```
nit: ambiguous identifiers - do the classes made the syntax more concise, etc or the facades

rustagir commented on docs/query-builder.txt:
```diff
@@ -9,588 +9,1095 @@ Query Builder
-**Where**
+{+odm-short+} aliases the ``DB`` method ``table()`` as the ``collection()``
+method. Chain methods to specify the command and any constraints. Then, chain
+the ``get()`` method at the end to run them on the MongoDB collection. The
```
```suggestion
the ``get()`` method at the end to run the command on the MongoDB collection. The
```

ccho-mongodb replied:
Thanks for spotting this! I'll reword since "them" refers to the unspecified number of chained methods.

rustagir commented on docs/query-builder.txt:
```diff
@@ -9,588 +9,1095 @@ Query Builder
-       Post::where('votes', '>', 0)
-           ->orWhere('is_approved', true)
-           ->get();
+This guide shows examples of the following types of query builder operations:
```
```suggestion
This guide provides examples for the following types of query builder operations:
```

ccho-mongodb replied:
I considered the suggestion while, but think that "for" might not be the correct preposition. I think "for" would imply that the examples are being delivered to someone or something. I'll use "provides" and pair it with "of".

rustagir commented on docs/query-builder.txt:
```diff
@@ -9,588 +9,1095 @@ Query Builder
+application, load sample datasets into your MongoDB deployment, and
+run the example code from a controller method.
+
+To perform read and write operations by using the query builder, import the
+``Illuminate\Support\Facades\DB`` facade and compose your query.
```
Q: would it be helpful to show a code block with the import statement?

ccho-mongodb replied:
Good question! Based on feedback in a prior PR from @GromNaN, the developer would be using an IDE which would auto-suggest the imports. In this case, although it might get auto-suggested, I think it's helpful to mention the import package to help the reader know where it came from.

rustagir commented on docs/query-builder.txt:
```diff
@@ -9,588 +9,1095 @@ Query Builder
+The following example shows how to chain the ``orWhere()``
+query builder method to retrieve documents from the
+``movies`` collection that either match the ``year``
+value of ``1955`` or match the title "``Back to the Future``":
```
```suggestion
value of ``1955`` or match the title ``"Back to the Future"``:
```

rustagir commented on docs/query-builder.txt:
```diff
@@ -9,588 +9,1095 @@ Query Builder
+   :language: php
+   :dedent:
+   :start-after: begin query orWhere
+   :end-before: end query orWhere
+
```
Q: is there no output for these examples, or did you purposely exclude the output box for brevity?

ccho-mongodb replied:
Good question! 
Yes, I purposely excluded it. In earlier drafts, I had started including the output for everything. However, I consciously decided to exclude output of the examples in which the output wouldn't contribute to building understanding since they would either have to know details about the dataset and/or the result is very similar to the description of the query.

rustagir commented on docs/query-builder.txt:
```diff
@@ -9,588 +9,1095 @@ Query Builder
+The following example shows how to chain the ``where()``
+query builder method to retrieve documents from the
+``movies`` collection that match both an ``imdb.rating``
+value greater than "8.5" and a ``year`` value of less than
```
```suggestion
value greater than ``8.5`` and a ``year`` value of less than
```

rustagir commented on docs/query-builder.txt:
```diff
@@ -9,588 +9,1095 @@ Query Builder
+The following example shows how to chain the ``where()``
+query builder method to retrieve documents from the
+``movies`` collection that match both an ``imdb.rating``
+value greater than ``8.5`` and a ``year`` value of less than
+``1940``:
+
```
I: this description does not match the code example
S: update the description to match or update the code to match

Might want to double check all instances - I will also be looking

ccho-mongodb replied:
Good catch! Copy-pasta from the previous example. Will check the rest again!

rustagir commented on docs/query-builder.txt:
```diff
@@ -9,588 +9,1095 @@ Query Builder
+~~~~~~~~~~~~~~~~~~~~~~
+
+The examples in this section show the query builder syntax you
+can use to perform to match values by using the following range
```
```suggestion
can use to match values by using the following range
```

rustagir commented on docs/query-builder.txt:
```diff
@@ -9,588 +9,1095 @@ Query Builder
 
-**whereIn**
+      [
+         { "title" "The Godfather", "imdb": { "rating": 9.2, "votes": 1038358, "id": 68646 }, ... },
+         { "title": "Hollywood", "imdb": { "rating": 9.1, "votes": 511,"id": 80230 }, ... },
+         { "title": "Cosmos", "imdb": { "rating": 9.3, "votes": 17174, "id": 81846 }, ... },
```
I: the text highlighting is a little off here. Is this a known issue or is there a different highlighting language we can use?

ccho-mongodb replied:
Thanks for catching that. I'll use "none" instead of JSON. I think it probably fails on the "..." as that's not valid JSON.

rustagir commented on docs/query-builder.txt:
```diff
@@ -9,588 +9,1095 @@ Query Builder
+   :language: php
+   :dedent:
+   :start-after: begin query whereNull
+   :end-before: end query whereNull
```
I: no output

ccho-mongodb replied:
Thanks for raising this. I purposely omitted this given that it's demonstrating omission and it probably raises questions as to why there's a missing field. Additional reasons for omitting output 
 listed here: https://github.com/mongodb/laravel-mongodb/pull/2790#discussion_r1538273062

rustagir commented on docs/query-builder.txt:
```diff
@@ -9,588 +9,1095 @@ Query Builder
+.. seealso::
 
-   $posts = Post::whereBetween('votes', [1, 100])->get();
+   The ``exists()`` MongoDB operation performs the same query.
```
Q: wouldnt the same function be negating the `exists` method?

ccho-mongodb replied:
Yes, that's correct. Good catch, will update.

rustagir commented on docs/query-builder.txt:
```diff
@@ -9,588 +9,1095 @@ Query Builder
+Aggregations
+~~~~~~~~~~~~
+
+The examples in this section show the query builder syntax you
+can use to perform aggregations, a set of calculations on the
+query results, to return the following information:
```
S: break up sentences

```suggestion
The examples in this section show the query builder syntax you
can use to perform aggregations. Aggregations are a set of calculations on the
query results that can help you to return the following information about your data:
```

ccho-mongodb replied:
Thanks for the suggestion. I'll break up the sentences to make it more clear.

rustagir commented on docs/query-builder.txt:
```diff
@@ -9,588 +9,1095 @@ Query Builder
+query results, to return the following information:
+
+- :ref:`Results grouped by common field values <laravel-query-builder-aggregation-groupby>`
+- :ref:`Count the of results <laravel-query-builder-aggregation-count>`
```
```suggestion
- :ref:`Count the number of results <laravel-query-builder-aggregation-count>`
```

ccho-mongodb replied:
Good catch!

rustagir commented on docs/query-builder.txt:
```diff
@@ -9,588 +9,1095 @@ Query Builder
+      :start-after: begin query groupBy
+      :end-before: end query groupBy
+
+   .. output::
+      :language: json
+      :visible: false
```
S: this example would be more illustrative if it produced a result where there were multiple titles per runtime.

ccho-mongodb replied:
Thanks for the suggestion! I think it doesn't work that way since it applies the $last aggregation operator to fields that are not part of the aggregated _id as mentioned in the tip admonition, but perhaps @GromNaN can suggest a solution?

rustagir commented on docs/query-builder.txt:
```diff
@@ -9,588 +9,1095 @@ Query Builder
+Aggregate Matched Results Example
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The following example shows how to use the ``sum()``
+query builder method to return the numerical total of
+the ``imdb.votes`` values from movies released in
+after the year "2000":
```
I: description does not match the example

ccho-mongodb replied:
Good catch, thanks!

rustagir commented on docs/query-builder.txt:
```diff
@@ -9,588 +9,1095 @@ Query Builder
+Aggregate Matched Results Example
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The following example shows how to use the ``sum()``
+query builder method to return the numerical total of
+the ``imdb.votes`` values from movies released in
+after the year "2000":
```
```suggestion
The following example shows how to use the ``sum()``
query builder method to return the numerical total of
the ``imdb.votes`` values from movies released in
after the year ``2000``:
```

rustagir commented on docs/query-builder.txt:
```diff
@@ -9,588 +9,1095 @@ Query Builder
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The following example shows how to use the ``skip()`` query builder method to
+omit the first "4" results that match the filter specified in the ``title``
```
```suggestion
omit the first 4 results that match the filter specified in the ``title``
```

ccho-mongodb replied:
Thanks for the suggestion! I'll use "four" as the quantifier since it's less than 10.

rustagir commented on docs/query-builder.txt:
```diff
@@ -9,588 +9,1095 @@ Query Builder
+
+The following example shows how to use the ``skip()`` query builder method to
+omit the first "4" results that match the filter specified in the ``title``
+field, sorted by the ``year`` value in descending order:
```
I; match example
```suggestion
field, sorted by the ``year`` value in ascending order:
```

ccho-mongodb replied:
Good catch!

rustagir commented on docs/includes/query-builder/MovieController.php:
```diff
@@ -0,0 +1,691 @@
+        // begin query all
+        $result = DB::collection('movies')
+            ->where('movies', 'all', ['title', 'rated', 'imdb.rating'])
+            ->getl();
```
```suggestion
            ->get();
```

rustagir commented on docs/query-builder.txt:
```diff
@@ -9,588 +9,1095 @@ Query Builder
+~~~~~~~~~~~~~~~~~~~~~~~
 
-.. code-block:: php
+The following example shows how to pass the ``type``
+query operator with the ``where()`` query builder
+method to match documents that contain a ``directors``
+field that contains an array of exactly five elements:
```
I: description and example do not demonstrate the functionality described in this section for type

ccho-mongodb replied:
Good catch, thanks!

rustagir commented on docs/query-builder.txt:
```diff
@@ -9,588 +9,1095 @@ Query Builder
+   :end-before: end query raw
 
-   User::where('age', 'exists', true)->get();
+The following code shows the equivalent MongoDB query:
```
```suggestion
The following code shows the equivalent MongoDB Shell query:
```

ccho-mongodb replied:
Thanks for the suggestion! I think it could be unexpected to introduce MongoDB Shell in passing, so I'll re-iterate that it is MongoDB Query API syntax.

rustagir commented on docs/query-builder.txt:
```diff
@@ -9,588 +9,1095 @@ Query Builder
+~~~~~~~~~~~~~~~~~~~~~~~
 
-   User::where('results', 'elemMatch', ['gte' => 80, 'lt' => 85])->get();
+The following example shows how to use the ``near`` query operator
+with the ``where()`` query builder method to match documents that contain the
+field ``random_review``:
```
I; description does not match

rustagir commented on docs/query-builder.txt:
```diff
@@ -9,588 +9,1095 @@ Query Builder
-**Push**
+MongoDB Write Operations
+------------------------
 
```
S: add introduction to section list

rustagir commented on docs/query-builder.txt:
```diff
@@ -9,588 +9,1095 @@ Query Builder
 
-If you **DON'T** want duplicate items, set the third parameter to ``true``:
+The following example shows how to use the ``increment()``
+query builder method to add "3000" to the value of
```
```suggestion
query builder method to add ``3000`` to the value of
```

rustagir commented on docs/query-builder.txt:
```diff
@@ -9,588 +9,1095 @@ Query Builder
-           'message' => 'Hi John',
-       ]);
+The following example shows how to use the ``push()`` query builder method to
+add "``Gary Cole``" to the ``cast`` array field in the matched document:
```
```suggestion
add ``"Gary Cole"`` to the ``cast`` array field in the matched document:
```

rustagir commented on docs/query-builder.txt:
```diff
@@ -9,588 +9,1095 @@ Query Builder
 
-.. code-block:: php
+The following example shows how to use the ``pull()`` query builder method
+to remove the "``Adventure``" value from the ``genres`` field from the document
```
```suggestion
to remove the ``"Adventure"`` value from the ``genres`` field from the document
```

rustagir reviewed (changes requested):
Good work on this long page! Happy to discuss any of my comments

Commit:
PRR fixes

Commit:
tweaks

Commit:
use unit test file

Commit:
reorder

Commit:
fixes

ccho-mongodb commented:
@rustagir thanks for the review. I think I addressed everything and went through the content multiple times. I also moved all the code snippets to a new unit test file.

rustagir commented on docs/query-builder.txt:
```diff
@@ -9,588 +9,1116 @@ Query Builder
-**whereNull**
+This guide provides examples of the following types of query builder operations:
+
+- :ref:`laravel-retrieve-query-builder`
+- :ref:`laravel-modify-results-query-builder`
+- :ref:`laravel-mongodb-read-query-builder`
+- :ref:`laravel-mongodb-write-query-builder`
```
S: Make the titles of the subsections all in the same form (action oriented)

rustagir commented on docs/query-builder.txt:
```diff
@@ -9,588 +9,1116 @@ Query Builder
+Where Method Example
+~~~~~~~~~~~~~~~~~~~~
+
+The following example shows how to use the ``where()`` query
+builder method to retrieve documents from the ``movies`` collection
+that contain an ``imdb.votes`` field value of ``350``. Click the
+:guilabel:`{+code-output-label+}` button to see the results returned
+by the query:
+
+.. io-code-block::
+   :copyable: true
+
+   .. input:: /includes/query-builder/QueryBuilderTest.php
+      :language: php
+      :dedent:
+      :start-after: begin query where
+      :end-before: end query where
```
I: description does not match example
imdb.votes field value of 350
where('imdb.rating', 9.3)

rustagir commented on docs/query-builder.txt:
```diff
@@ -9,588 +9,1116 @@ Query Builder
+The following example shows how to chain the ``orWhere()``
+query builder method to retrieve documents from the
+``movies`` collection that either match the ``year``
+value of ``1955`` or match the title ``"Back to the Future"``:
```
```suggestion
value of ``1955`` or match the ``title`` value ``"Back to the Future"``:
```

rustagir commented on docs/query-builder.txt:
```diff
@@ -9,588 +9,1116 @@ Query Builder
+The following example shows how to use the ``whereIn()``
+query builder method to retrieve documents from the
+``movies`` collection that match at least one of the
+``title`` values in the specified set.
```
```suggestion
``title`` values in the specified set:
```

rustagir commented on docs/query-builder.txt:
```diff
@@ -9,588 +9,1116 @@ Query Builder
+The following example shows how to use the ``whereNull()``
+query builder method to retrieve documents from the
+``movies`` collection that omit a ``runtime`` value
+or field.
```
```suggestion
or field:
```

rustagir commented on docs/query-builder.txt:
```diff
@@ -9,588 +9,1116 @@ Query Builder
+wildcard characters:
+
+- ``%`` which matches zero or more characters
+- ``_`` which matches a single character
```
Q: if the `_` matches a single character, why is Spider Woman in the results?

ccho-mongodb replied:
> Q: if the `_` matches a single character, why is Spider Woman in the results?

Good question -- whitespace is a character in the context of SQL queries. The [Laravel documentation](https://laravel.com/docs/11.x/queries#where-any-all-clauses) omits any explanation of the wildcard characters that they use, but I thought it would be good to reiterate the SQL LIKE syntax.
Let me know if you think this needs further explanation.

rustagir commented on docs/query-builder.txt:
```diff
@@ -9,588 +9,1116 @@ Query Builder
+The following example shows how to pass the ``mod``
+query operator with the ``where()`` query builder
+method to match documents by using the expression
+``"year % 2 == 0"``, which matches even values for
```
S: remove quotes from the expression 
```suggestion
``year % 2 == 0``, which matches even values for
```

rustagir reviewed (approved):
lgtm with some small fixes

Commit:
PRR fixes

Commit:
PRR fix 2

Commit:
grammar fix

GromNaN commented on docs/query-builder.txt:
```diff
@@ -9,588 +9,1116 @@ Query Builder
+~~~~~~~~~~~~~~~~~~~~~~~
 
-**Type**
+The following example shows how to use the ``near`` query operator
+with the ``where()`` query builder method to match documents that
+contain a location that is up to ``50`` meters from a GeoJSON Point
+object:
```
This should tell how to create a geo index.

ccho-mongodb replied:
Will do -- I can link to the Schema Builder section "Create a Geospatial Index". However, I'll need to merge in the 4.1 branch to get that content. Is that ok or should I create a separate ticket to address this?

GromNaN commented on docs/includes/query-builder/QueryBuilderTest.php:
```diff
@@ -0,0 +1,542 @@
+    {
+        // <optionally, add code here to clean the database/collection>
+        // begin query near
+        $results = DB::collection('theaters')
```
There is a mix of `$result` and `$results` (with s) in the examples. It would be better to remain consistent with `$result` (without s)

ccho-mongodb replied:
We usually use plural variable names to distinguish between single and multiple results expected. Do you feel strongly about using only $result? If so, we can change our convention for docs in this repo -- could we make that a followup task so we can revisit them all at once?

GromNaN replied:
I wanted to make a note of it to make sure it was intentional. I leave this decision up to you.

GromNaN committed:
Update tests with data

GromNaN reviewed (commented):
I updated the tests to add assertions, using a subset of the data.

Commit:
Merge branch '4.1' into DOCSP-35966-query-builder

Commit:
add link to schema builder for info on creating geospatial indexes
